src/py21cmfast/io/h5.py

```diff
@@ -339,23 +350,51 @@ def _read_outputs(group: h5py.Group):
             stacklevel=2,
         )
     else:
-        return _read_outputs_v4(group)
+        return _read_outputs_v4(group, struct, redshift, inputs)
 
 
-def _read_outputs_v4(group: h5py.Group):
-    arrays = {
-        name: Array(
-            dtype=box.dtype,
-            shape=box.shape,
-            state=ArrayState(on_disk=True),
-            cache_backend=H5Backend(path=group.file.filename, dataset=box.name),
-        )
-        for name, box in group.items()
-    }
-    for k, val in group.attrs.items():
-        if k == "21cmFAST-version":
-            continue
+def _read_outputs_v4(
+    group: h5py.Group,
+    struct: type[ostruct.OutputStruct],
+    redshift: float | None,
+    inputs: InputParameters,
+):
+    # First read other attributes that are not arrays.
+    kwargs = dict(group.attrs)
+    del kwargs["21cmFAST-version"]
 
-        arrays[k] = val
+    if redshift is not None:
+        kwargs["redshift"] = redshift
+
+    # Create the object with those attributes.
+    obj = struct.new(inputs, **kwargs)
+
+    # Now go and make sure all the arrays exist in the file, and have the correct shape.
+    # We don't actually read these right now, we just make pointers to the file.
+    for name, array in obj.arrays.items():
+        if name not in group:
+            raise HDF5FileStructureError(
+                f"Required Array {name} not found in {group}. This file is not valid."
+            )
+
+        dataset = group[name]
+        if dataset.shape != array.shape:
+            raise HDF5FileStructureError(
+                f"Array {name} has shape {dataset.shape} in the file {group.file.filename}, but requires shape {array.shape}"
+            )
+
+        # We don't check dtype because it can be usually safely cast.
+
+        setattr(
+            obj,
+            name,
+            attrs.evolve(
+                array,
+                state=ArrayState(on_disk=True),
+                cache_backend=H5Backend(
+                    path=group.file.filename, dataset=group.name + f"/{name}"
+                ),
+            ),
+        )
```

Comment:
(comment missed in review) I spent a while trying to figure out whether the given file could be accidentally overwritten by the `Array` module here on a compute. Since the caching explicity passes a new `CacheBackend` to `Array.written_to_disk` the only way this could happen is if a user calls `Array.written_to_disk`, after explicitly calling `.compute(allow_already_computed=True) with `write=False`, which I don't think will ever happen.